common/nexus/incoming_service_registry.go

```diff
@@ -0,0 +1,355 @@
+		tableVersion int64
+		services     map[string]*nexus.IncomingService // mapping of service ID -> service
+
+		refreshPoller *goro.Handle
+
+		matchingClient matchingservice.MatchingServiceClient
+		persistence    p.NexusIncomingServiceManager
+		logger         log.Logger
+	}
+)
+
+func NewIncomingServiceRegistryConfig(dc *dynamicconfig.Collection) *IncomingServiceRegistryConfig {
+	config := &IncomingServiceRegistryConfig{
+		refreshLongPollTimeout: dc.GetDurationProperty(dynamicconfig.FrontendRefreshNexusIncomingServicesLongPollTimeout, 5*time.Minute),
+		refreshPageSize:        dc.GetIntProperty(dynamicconfig.NexusIncomingServiceListDefaultPageSize, 1000),
+		refreshMinWait:         dc.GetDurationProperty(dynamicconfig.FrontendRefreshNexusIncomingServicesLongPollTimeout, 1*time.Second),
+	}
+	config.refreshRetryPolicy = backoff.NewExponentialRetryPolicy(config.refreshMinWait()).WithMaximumInterval(config.refreshLongPollTimeout())
+	return config
+}
+
+func NewIncomingServiceRegistry(
+	config *IncomingServiceRegistryConfig,
+	matchingClient matchingservice.MatchingServiceClient,
+	persistence p.NexusIncomingServiceManager,
+	logger log.Logger,
+) *IncomingServiceRegistry {
+	return &IncomingServiceRegistry{
+		status:           common.DaemonStatusInitialized,
+		config:           config,
+		serviceDataReady: future.NewFuture[struct{}](),
+		services:         make(map[string]*nexus.IncomingService),
+		matchingClient:   matchingClient,
+		persistence:      persistence,
+		logger:           logger,
+	}
+}
+
+func (r *IncomingServiceRegistry) Start() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	refreshCtx := headers.SetCallerInfo(
+		context.Background(),
+		headers.SystemBackgroundCallerInfo,
+	)
+	r.refreshPoller = goro.NewHandle(refreshCtx).Go(r.refreshServicesLoop)
+}
+
+func (r *IncomingServiceRegistry) Stop() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	r.refreshPoller.Cancel()
+	<-r.refreshPoller.Done()
+}
+
+func (r *IncomingServiceRegistry) Get(ctx context.Context, id string) (*nexus.IncomingService, error) {
+	if !r.serviceDataReady.Ready() {
+		if err := r.initializeServices(ctx); err != nil {
+			r.logger.Error("error initializing Nexus incoming service cache.", tag.Error(err))
+			return nil, serviceerror.NewInternal("error loading Nexus incoming service data.")
+		}
+	}
+
+	r.dataLock.RLock()
+	defer r.dataLock.RUnlock()
+
+	service, ok := r.services[id]
+	if !ok {
+		return nil, serviceerror.NewNotFound(fmt.Sprintf("could not find Nexus incoming service with ID: %v", id))
+	}
+
+	return service, nil
+}
+
+func (r *IncomingServiceRegistry) refreshServicesLoop(ctx context.Context) error {
+	minWaitTime := r.config.refreshMinWait()
+
+	for ctx.Err() == nil {
+		start := time.Now()
+		_ = backoff.ThrottleRetryContext(ctx, r.refreshServices, r.config.refreshRetryPolicy, nil)
```

Comment:
Not sure why you're ignoring the error here.
Can you add a comment?

Reply:
Not sure what we would do with that error other than maybe log it (but I think we have logging for all the relevant errors closer to where they occur).

The logic is the same as https://github.com/temporalio/temporal/blob/main/service/matching/task_queue_manager.go#L1074

Reply:
Seems okay to me, I'd just add a comment explaining the desired behavior and reasoning.